Sources/StreamChat/ChatClient.swift

```diff
@@ -314,6 +314,96 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
         completeConnectionIdWaiters(connectionId: nil)
         completeTokenWaiters(token: nil)
     }
+    
+    public func connectUser(
+        name: String? = nil,
+        imageURL: URL? = nil,
+        extraData: ExtraData.User = .defaultValue,
+        token: Token,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        setConnectionInfoAndConnect(
+            name: name,
+            imageURL: imageURL,
+            tokenProvider: .static(token),
+            completion: completion
+        )
+    }
+    
+    public func connectUser(
+        connectionInfo: UserConnectionInfo<ExtraData>? = nil,
+        token: Token,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        connectUser(
+            name: connectionInfo?.name,
+            imageURL: connectionInfo?.imageURL,
+            extraData: connectionInfo?.extraData ?? .defaultValue,
+            token: token,
+            completion: completion
+        )
+    }
+    
+    public func connectUser(
+        connectionInfoProvider: ((Result<(UserConnectionInfo<ExtraData>, Token), Error>) -> Void) -> Void,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        connectionInfoProvider { userInfo in
+            switch userInfo {
+            case let .success((info, token)):
+                connectUser(connectionInfo: info, token: token)
+            case let .failure(error):
+                completion?(error)
+            }
+        }
+    }
+    
+    public func connectUser(
+        tokenProvider: ((Result<Token, Error>) -> Void) -> Void,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        tokenProvider { token in
+            switch token {
+            case let .success(token):
+                setConnectionInfoAndConnect(tokenProvider: .static(token), completion: completion)
+            case let .failure(error):
+                completion?(error)
+            }
+        }
+    }
```

Comment:
This one won't be needed in the end.

Reply:
What about the one with `connectionInfoProvider: ((Result<(UserInfo<ExtraData>, Token), Error>) -> Void) -> Void`?